VM Upgrade: T.S. '89

NullVoxPopuli opened the pull request:
![image](https://github.com/emberjs/ember.js/assets/199018/ba3bf9e5-d17f-45ea-8cb1-5ee7a945a659)

Changelog: https://github.com/glimmerjs/glimmer-vm/releases/tag/v0.89.0


I had some issues with `this.attrs`, so I just remove the problematic tests, since `attrs` in general was supposed to be removed in v4, see: https://github.com/emberjs/ember.js/pull/20671 for that effort.

NullVoxPopuli committed:
T.S. '89

NullVoxPopuli committed:
Handle VM breaking changes (new info for modifiers, updated AST)

NullVoxPopuli committed:
Use non-deprecated AST Nodes

NullVoxPopuli committed:
@babel/plugin-proposal-class-properties is required to build ember

NullVoxPopuli committed:
More plugins updates

NullVoxPopuli committed:
'89 -> '90

NullVoxPopuli committed:
'90 -> '91

NullVoxPopuli committed:
New debug-render-tree-nodes to add to existing tests

NullVoxPopuli committed:
Very wip -- something may have gone wrong with the AST stuff in the VM

NullVoxPopuli committed:
shrug

NullVoxPopuli committed:
Finish updating debug-render-tree tests

NullVoxPopuli committed:
TODO: need to move assert-splattribute-expression to glimmer

NullVoxPopuli committed:
Move assert test for splattributes in {{ }} that is now a hard error from @glimmer/syntax

NullVoxPopuli committed:
0.90.1 -> 0.91.0

NullVoxPopuli committed:
Fix mistaken find and replace

NullVoxPopuli committed:
Fix the state / scope tracker

NullVoxPopuli committed:
Remove mistaken plugin from strict-mode transforms

NullVoxPopuli committed:
Remove tests for code that was supposed to be removed in v4

NullVoxPopuli committed:
Remove extraneous import

NullVoxPopuli committed:
0.91.0 -> 0.91.1

chancancode commented:
Rebase after https://github.com/emberjs/ember.js/pull/20672 lands

chancancode commented on packages/@ember/-internals/glimmer/lib/modifiers/action.ts:
```diff
@@ -200,6 +200,10 @@ class ActionModifierManager implements InternalModifierManager<ActionState, obje
 
+  getDebugInstance(state: ActionState): unknown {
+    return state;
+  }
```
I think we talked about this and we want to return `null`? We don't really want to leak the internal state bucket, and it's not really useful for it to show up on the inspector

NullVoxPopuli replied:
right! updated

chancancode commented on packages/@ember/-internals/glimmer/lib/modifiers/internal.ts:
```diff
@@ -78,6 +78,10 @@ export class InternalModifierManager
 
+  getDebugInstance(state: InternalModifierState): unknown {
+    return state;
+  }
```
I think just declare the class `abstract` and leave this out. We have to decide case-by-case what we want to expose here. And so far this is only used by `{{action}}` anyway

NullVoxPopuli replied:
TS says its required, -- I'll return null by default tho

chancancode replied:
Yeah, that's why I said *declare the class `abstract`*... when you _concretely_ implement/extend an abstract class, you are required to implement all the abstract members (which is the point of abstract classes). 

_However,_ you can also abstractly extend a class (`abstract class Foo extends AnotherAbstractClass`), in which case, you are allowed to leave out any abstract members, essentially deferring/propagating the requirement to your children

chancancode commented on packages/@ember/-internals/glimmer/tests/integration/components/curly-components-test.js:
```diff
@@ -2492,7 +2492,7 @@ moduleFor(
           value: null,
         }),
-        template: '<div id="inner-value">{{wrapper.content}}</div>',
+        template: '<div id="inner-value">{{this.wrapper.content}}</div>',
```
Something is fishy/scary that these continued to work, seemingly without assertions/deprecations either. Hopefully the new VM code is erroring consistently now, but this means we may have implicit this code in the wild that was accidentally working

kategengler replied:
I think `expectAssertion` below would have swallowed any other assertions, even if it matched on `backtrackingMessageFor`

chancancode commented on packages/@ember/-internals/glimmer/tests/integration/helpers/hash-test.js:
```diff
@@ -217,92 +217,5 @@ moduleFor(
-      });
-
-      this.assertText('Godfrey Chan');
-    }
```
This went from deprecation to nothing (instead of an assertion). I don't think it's illegal per-se, I think we are technically allowed to do it, but that's when we tend to get ourselves into trouble – see the `{{#with}}` thing recently. Next time at least consider whether we can keep an assertion around for at least 1-2 LTS.

chancancode commented on packages/ember-template-compiler/lib/plugins/assert-against-attrs.ts:
```diff
@@ -81,9 +81,8 @@ function isAttrs(node: AST.PathExpression, symbols: string[]) {
-      node.parts.shift();
-      node.original = node.original.slice(5);
+    if (node.head.type === 'ThisHead') {
+      node.tail.shift();
```
Rebase away

NullVoxPopuli replied:
other PR isn't merged yet :sweat_smile:

chancancode replied:
Merged now!

NullVoxPopuli replied:
updated!

chancancode commented on packages/ember-template-compiler/lib/plugins/transform-resolutions.ts:
```diff
@@ -149,7 +149,9 @@ export default function transformResolutions(env: EmberASTPluginEnvironment): AS
 function isLocalVariable(node: AST.PathExpression, hasLocal: (k: string) => boolean): boolean {
-  return !node.this && node.parts.length === 1 && hasLocal(node.parts[0]!);
+  return (
+    !(node.head.type === 'ThisHead') && node.tail.length === 1 && hasLocal(node.head.original!)
```
Why the ! on `.original`? It should always exist?

NullVoxPopuli replied:
yes, but nothing complained about it. It's been removed now.

chancancode commented on packages/ember-template-compiler/lib/plugins/transform-wrap-mount-and-outlet.ts:
```diff
@@ -43,8 +43,10 @@ export default function transformWrapMountAndOutlet(env: EmberASTPluginEnvironme
-      Program: node,
+      Template: node,
       ElementNode: node,
+      BlockStatement: node,
```
this is wrong (see below)

NullVoxPopuli replied:
thanks! updated!

chancancode commented on packages/ember-template-compiler/lib/plugins/utils.ts:
```diff
@@ -12,25 +12,41 @@ export function isStringLiteral(node: AST.Expression): node is AST.StringLiteral
   let node = {
-    enter(node: AST.Program | AST.Block | AST.ElementNode) {
-      for (let param of node.blockParams) {
+    enter(node: AST.Template | AST.Block | AST.ElementNode | AST.BlockStatement) {
```
You don't want `AST.BlockStatement` here. `BlockStatement` is the `{{#foo}}...{{/foo}}`, the block params is on the default/inverse blocks (which should get visited by the `Block` visitor). 

There _is_ a `params` on `BlockStatement` but it means a completely different thing, it is the positional argument for the invocation, which has a type of `AST.Expression[]` hence you question above.

On the other hand, `node.params` on `AST.Template | AST.Block | AST.ElementNode` should all be `AST.VarHead[]`.

chancancode commented on packages/ember-template-compiler/lib/plugins/utils.ts:
```diff
@@ -12,25 +12,41 @@ export function isStringLiteral(node: AST.Expression): node is AST.StringLiteral
-    enter(node: AST.Program | AST.Block | AST.ElementNode) {
-      for (let param of node.blockParams) {
+    enter(node: AST.Template | AST.Block | AST.ElementNode | AST.BlockStatement) {
+      let params = 'params' in node ? node.params : node.blockParams;
```
All the nodes you checked here should have `params`. Why this check? Unless you intend for this `enter` hook to be more generic, in which case you need to type `node: AST.Node` and do a different check here, because there are other things with a `params` property that is unrelated.

NullVoxPopuli replied:
`params does not exist on type Template` - we can play with removing it I guess, but I feel like it was needed for something and I don't remember -- I'll see if I rediscover that tomorrow

chancancode replied:
oh yeah, that’s right, cause we don’t have locs for the js scope locals 👍

chancancode commented on packages/ember-template-compiler/lib/plugins/utils.ts:
```diff
@@ -12,25 +12,41 @@ export function isStringLiteral(node: AST.Expression): node is AST.StringLiteral
+    enter(node: AST.Template | AST.Block | AST.ElementNode | AST.BlockStatement) {
+      let params = 'params' in node ? node.params : node.blockParams;
+      for (let param of params) {
+        let name = getLocalName(param);
```
Once you fix the types/guards `getLocalName` shouldn't need to exist

NullVoxPopuli replied:
Still needs to exist because `param` can be of type string (from `blockParams`), which exist on `Template`

chancancode commented on packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js:
```diff
@@ -1,48 +0,0 @@
-import { moduleFor, AbstractTestCase } from 'internal-test-helpers';
```
I think we should keep this around and just change to use `assert.throws`. Generally we want to keep the coverage even if it duplicates what is already tested in the VM. The rule of thumb is if some day we decided to replace `glimmer-vm` with something else, we shouldn't loose coverage on anything that ember cares about.

NullVoxPopuli replied:
`@glimmer/syntax` would still be used tho, even if we swap out the VM

NullVoxPopuli replied:
(I'm intending on leaving this file deleted, partly because `@glimmer/syntax` is not glimmer-vm)

chancancode replied:
Not necessarily, all those packages were invented out of thin air at the same time, it’s still an external (or internal, I guess?) dependency and ember needs to test its public api

NullVoxPopuli replied:
It has it's own tests to tho. And we're planning on moving all that in to this repo

chancancode replied:
Yes, that's what we have _always_ done. I didn't invent it just to slow down this particular PR. Often we have almost exactly the same test twice. See, for example: https://github.com/emberjs/ember.js/pull/17499

The tests are already there and it should be easy enough to change them to match (it's ok to not match the error messages very precisely, since the point is just to check that there are _some_ errors. 

Could we revisit how we do thing in general? Yes? But as you said you don't also want the creep the scope of the PR so unless this is a particularly important hill for you we can discuss that another time/day.

If it did turn out to be very challenging to keep the tests working we can of course revisit if it's worth the time. But as a matter of default, this is the status quo (and I don't think it's a pointless status quo/just useless formalities), and is what we should do.

NullVoxPopuli replied:
The intent of this file has been restored, albeit, rewritten due to the nature of how the error was surfaced.

I removed the line checking, because, by the logic given, it's not relevant, and we want to make sure that syntax restrictions don't regress, due to not trusting `@glimmer/syntax` :sweat_smile: (which I don't agree with, but I can't care anymore :shrug: <3 )

chancancode commented on types/publish.mjs:
```diff
@@ -109,8 +109,7 @@ async function main() {
   // we still need to post-process them.
-  let excludes = remappedLocationExcludes
-    .concat(sideEffectExcludes);
+  let excludes = remappedLocationExcludes.concat(sideEffectExcludes);
```
This seems unrelated, it doesn't hurt but I'd revert it just to keep the diff minimal. I did notice we don't really have prettier set up here, which we can add in a different PR if we want.

NullVoxPopuli replied:
yeah, it's annoying to have an editor that enables prettier when a `.prettierrc.js` is detected, but some files haven't been formatted with it. :(

NullVoxPopuli replied:
updated

chancancode commented on packages/ember-template-compiler/tests/plugins/transform-component-invocation-test.js:
I don't know what this test file is trying to test anymore 😐

This probably corresponds to some kind of transform we did here to make some of these syntax work before they were implemented in the VM (this file probably predates the VM even existing), but we don't do that anymore, and the test doesn't actually do anything other than, checking these syntax compiles without error. 

Surely I hope that is already covered indirectly by any tests that actually verifies/exercises the feature. This is not that different from smoke testing `<p>hi</p>` and `{{this.foo}}` compiles without error, which, I suppose isn't inherently useless, but if we are going to do that we would want to add a more cases.

Not really your problem but I wonder if we should just delete this whole file.

NullVoxPopuli committed:
Do not expose modifier manager states

NullVoxPopuli committed:
Remove a !

NullVoxPopuli committed:
Revert types/publish.mjs

NullVoxPopuli committed:
Cleanup some types, don't use BlockStatement for scope tracking

NullVoxPopuli committed:
Fix old test: wrapper was not in scope

NullVoxPopuli committed:
restore/rewrite assert-splattribute-expression-test

NullVoxPopuli commented on packages/ember-template-compiler/tests/plugins/assert-splattribute-expression-test.js:
```diff
@@ -1,48 +1,27 @@
+      assert.throws(() => compile('<div>{{...attributes}}</div>'), message);
     }
 
-    '@test {{...attributes}} is not valid path expression'() {
```
I deleted this test entirely because it was the exact same as the one above

chancancode commented on packages/@ember/-internals/glimmer/lib/modifiers/internal.ts:
```diff
@@ -78,6 +78,10 @@ export class InternalModifierManager
     return null;
   }
 
+  getDebugInstance(): unknown {
+    return null;
+  }
+
```
```suggestion
```

chancancode replied:
Was hoping I could use this to commit the change directly from the UI, saving a roundtrip, but it wouldn't let me also add a diff for the `abstract class ...` because that line was unchanged 😐

NullVoxPopuli replied:
this is a required method

```
Class 'InternalModifierManager' incorrectly implements interface 'InternalModifierManager<InternalModifierState, typeof InternalModifier>'.
     Property 'getDebugInstance' is missing in type 'InternalModifierManager' but required in type 'InternalModifierManager<InternalModifierState, typeof InternalModifier>'. [2420]
```

NullVoxPopuli replied:
nevermind: https://github.com/emberjs/ember.js/pull/20658/commits/a13c6275687d119a8de0dd9cb0f9755df2b8ba4d#r1544906209

chancancode committed:
Remove default implementation of `getDebugInstance`

We have to make a judgement call per occurance to determine what we
should return here.

This whole mess could use some simplification, but we probably just
won't end up keeping this around for much longer once/if we remove
the action modifier, which is currently the only kind of modifier
we have internally (since `{{on}}` is implemented in glimmer-vm and
essentially duplicates this whole "internal" set up once more time
over there).

chancancode committed:
Remove outdated comment

NullVoxPopuli commented on packages/@ember/-internals/glimmer/lib/modifiers/internal.ts:
```diff
@@ -78,9 +78,7 @@ export class InternalModifierManager
-  getDebugInstance(): unknown {
-    return null;
-  }
+  abstract getDebugInstance(state: InternalModifierState): unknown;
```
ah, this is what you meant

chancancode replied:
Yeah, it’s a little bit of an edge case because we are implementing an interface. If we are extending an abstract super class you wouldn’t have to explicitly propagate it like this.

chancancode commented on packages/ember-template-compiler/lib/plugins/index.ts:
```diff
@@ -36,8 +34,6 @@ export const STRICT_MODE_TRANSFORMS = Object.freeze(
     TransformEachInIntoEach,
     TransformInElement,
-    AssertSplattributeExpressions,
-    TransformEachTrackArray,
```
This is revealing a real issue: https://github.com/emberjs/ember.js/pull/20673

We have to add it back, and deal it somehow on the glimmer-vm side. My initial thought is that glimmer-vm can assume everything that starts with a `-` is a keyword that ember will provide at runtime and not error on those (since it couldn't have been a JavaScript identifier), but maybe that will mask too many real errors when people are migrating existing templates (not uncommon to have helpers starting with dash in apps).

Maybe we'll have to add a `KeywordHead` subnode, or just make it something you are very unlikely to type, like `__GLIMMER_KEYWORD__` or something. Not sure.

But it is a real problem. If we don't deal with it, we would have broken `{{outlet}}`, `{{mount}}`, {{#each-in}}`, `{{#in-element}}` (perhaps among other things) for strict mode templates

chancancode committed:
add back `TransformEachTrackArray`

chancancode committed:
Re-apply Program -> Template fix after rebase

chancancode committed:
Missing Block visitor

chancancode committed:
Additional coverage for strict mode keywords

chancancode committed:
Pass list of strict mode keywords when calling precompile

NullVoxPopuli committed:
0.91.1 -> 0.91.2

NullVoxPopuli committed:
0.91.2 -> 0.92.0

NullVoxPopuli committed:
Update tests to accoutn for the (action) and {{action}} deprecation
